File: microPython/DFRobot_MAX31855.py
'''!
  @file DFRobot_MAX31855.py
  @brief A digital K type high temperature sensor library
  @copyright   Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
  @license     The MIT License (MIT)
  @author      PengKaixing(kaixing.peng@dfrobot.com)
  @version  V1.0.0
  @date  2022-04-01
  @url https://github.com/DFRobot/DFRobot_MAX31855
'''

class DFRobot_MAX31855:
  '''
    @brief A class to get sensor temperature
  '''
  def __init__(self,i2c):
    self.i2c = i2c
    
  def read_celsius(self):
    '''
      @brief A class to get sensor temperature
      @return Temperature value
    '''
    a,b,d = self.read_data()
    if(d&0x7):
      return None
    if(a&0x80):
      a = 0xff - a
      b = 0xff - b
      temp = -((((a << 8) | b) >> 2)+1)*0.25
      return temp
    temp = (((a << 8) | b) >> 2)*0.25
    return temp

  def read_data(self):
    '''
      @brief Read sensor data
      @return Sensor data
    '''
    i = 0
    a = self.i2c.readfrom_mem(16,0x00,1)[0]
    b = self.i2c.readfrom_mem(16,0x01,1)[0]
    d = self.i2c.readfrom_mem(16,0x03,1)[0]
    return a,b,d

File: DFRobot_MAX31855.cpp
/*!
 * @file DFRobot_MAX31855.cpp
 * @brief A digital K type high temperature sensor library
 * @copyright   Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
 * @license     The MIT License (MIT)
 * @author      PengKaixing(kaixing.peng@dfrobot.com)
 * @version  V1.0.0
 * @date  2022-04-01
 * @url https://github.com/DFRobot/DFRobot_MAX31855
 */

#include <DFRobot_MAX31855.h>

DFRobot_MAX31855::DFRobot_MAX31855(TwoWire *pWire, uint8_t I2C_addr)
{
  _pWire = pWire;
  _I2C_addr = I2C_addr;
}

void DFRobot_MAX31855::begin(void)
{
  _pWire->begin();
}

float DFRobot_MAX31855::readCelsius(void)
{
  unsigned char rxbuf[4] = {0};
  readData(0x00, rxbuf,4);
  if(rxbuf[3]&0x7){
    return NAN;
  }
  if(rxbuf[0]&0x80){
    rxbuf[0] = 0xff - rxbuf[0];
    rxbuf[1] = 0xff - rxbuf[1];
    float temp =  -((((rxbuf[0] << 8)|(rxbuf[1] & 0xfc)) >> 2) + 1) * 0.25;
    return temp;
  }
  float temp =(((rxbuf[0] << 8 )| (rxbuf[1] & 0xfc)) >> 2)*0.25;
  return temp;
}

int16_t DFRobot_MAX31855::readData(uint8_t Reg, uint8_t *Data, uint8_t len)
{
  int i = 0;
  _pWire->beginTransmission(this->_I2C_addr);
  _pWire->write(Reg);
  if (_pWire->endTransmission() != 0)
  {
    return -1;
  }
  _pWire->requestFrom((uint8_t)this->_I2C_addr, (uint8_t)len);
  while (_pWire->available())
  {
    Data[i++] = _pWire->read();
  }
  return len;
}
